runtime/src/serde_snapshot.rs

```diff
@@ -0,0 +1,761 @@
+    let (slot, bank_hash): (Slot, BankHashInfo) = deserialize_from(&mut stream)
+        .map_err(|e| format!("bank hashes deserialize error: {}", e.to_string()))
+        .map_err(accountsdb_to_io_error)?;
+    accounts_db
+        .bank_hashes
+        .write()
+        .unwrap()
+        .insert(slot, bank_hash);
+
+    // lastly generate accounts database index
+    accounts_db.generate_index();
+
+    Ok(accounts_db)
+}
+
+struct SerializableAccountsDatabaseUnversioned<'a, 'b, C> {
+    accounts_db: &'a AccountsDB,
+    slot: Slot,
+    account_storage_entries: &'b [SnapshotStorage],
+    phantom: std::marker::PhantomData<C>,
+}
+
+impl<'a, 'b, C: TypeContext<'b>> Serialize for SerializableAccountsDatabaseUnversioned<'a, 'b, C> {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        // sample write version before serializing storage entries
+        let version = self.accounts_db.write_version.load(Ordering::Relaxed);
+
+        // accounts database serializes as a 3-tuple
+        let mut seq = serializer.serialize_tuple(C::legacy_or_zero(1) + 3)?;
+
+        // possibly write out a dummy vector length for backward compatibility
+        C::legacy_serialize_byte_length(&mut seq, MAX_ACCOUNTS_DB_STREAM_SIZE)?;
+
+        // (1st of 3 elements) write the list of account storage entry lists out as a map
+        {
+            let mut serialize_account_storage_timer =
+                Measure::start("serialize_account_storage_ms");
+            let mut entry_count = 0;
+            seq.serialize_element(&serialize_iter_as_map(
+                self.account_storage_entries.iter().map(|x| {
+                    entry_count += x.len();
+                    (
+                        x.first().unwrap().slot,
+                        serialize_iter_as_seq(
+                            x.iter()
+                                .map(|x| C::SerializableAccountStorageEntry::from(x.as_ref())),
+                        ),
+                    )
+                }),
+            ))?;
+            serialize_account_storage_timer.stop();
+            datapoint_info!(
+                "serialize_account_storage_ms",
+                ("duration", serialize_account_storage_timer.as_ms(), i64),
+                ("num_entries", entry_count, i64),
+            );
+        }
+
+        // (2nd of 3 elements) write the current write version sampled before
+        // the account storage entries were written out
+        seq.serialize_element(&version)?;
+
+        // (3rd of 3 elements) write out (slot, bank hashes) pair
+        seq.serialize_element(&(
+            self.slot,
+            &*self
+                .accounts_db
+                .bank_hashes
+                .read()
+                .unwrap()
+                .get(&self.slot)
+                .unwrap_or_else(|| panic!("No bank_hashes entry for slot {}", self.slot)),
+        ))?;
+
+        seq.end()
```

Comment:
@svenski123
I appreciate what you've done for us. That's really impressive and I've learned advanced stuff for serde from you. For example, I've never come up with generic type parameterized versioned de-/seerialization scheme. I really like this part!

But then, we'd like more simplified approach to the solution.

I'm not sure NEWER/OLDER Future/Lagecy makes life easier for us. I think it will require us to rename this new `Future` to `Legacy` when we add another snapshot version? That's a bit tedious.

Instead, I think it's fine to embed explicit specific version numbers in constants. Even at a time, there might temporarily be 3 versions, that's fine.

In addition, we also want to avoid to define like `serialize_iter_as_seq`/`serialize_iter_as_map`/`serialize_iter_as_tuple` by ourselves unless it's absolutely needed.

And as [I commented yesterday here](https://github.com/solana-labs/solana/pull/9980#discussion_r427892754), I think there is more simplified approach.

For example, the code this comment is attached to is serializing AccountsDB. And the corresponding code in my suggested commit is like [this](https://github.com/ryoqun/solana/commit/6d58b73c47294bfb93465d5a83cd2175660b6e6d#diff-90c69309ec8cec7cf741c70e18bf2e86R276).

I think my suggestion is simpler and more manageable in my opinion.

Or I might be missing some. Could you explain why you think this pr's current approach is expected to be simper and future-proof?

Specifically, could you explain how could we reach to [this state](https://github.com/ryoqun/solana/commit/6d58b73c47294bfb93465d5a83cd2175660b6e6d) or even simpler versioned code starting from the current implementation?

I mean, could you elaborate a bit on this?: https://github.com/solana-labs/solana/pull/9980#discussion_r428278811

> Once the legacy serialisation style is no longer needed, much of this code can and should be removed, as well as the generic type parameterized implementation if it proves to be more trouble than it's worth (I must admit it is quite awkward, especially in rust).

Also, how could we remove the Legacy serialization debt along the course from now, including like `legacy_serialize_byte_length`, which is in the common codepath? Maybe just remove after we remove the current Legacy? If so, why cannot we go one step further and remove them now?

Reply:
###
@ryoqun Having pulled down your repo and looked at your commit in an editor (as opposed to github), your intent is much clearer to my now and much of my concern is proven unfounded. Obviously I have only myself to blame for this, so apologies. As I have written the comments below I will still share them in the hope that they may be useful, though please bear in mind that until close to the end I though you had a much less nuanced approach.
-Kris
###

> @svenski123
> I appreciate what you've done for us. That's really impressive and I've learned advanced stuff for serde from you. For example, I've never come up with generic type parameterized versioned de-/seerialization scheme. I really like this part!

It can help when you're faced with a large chunk of unfamiliar, complex code - cordon off / parameterize the piece you need to change, test it with the original parameterize to ensure the logic is maintained, then design a new type/piece to slot in knowing that a) the original behavior remains unaffected and b) any problems with the new behaviour are due solely to issues with the new type/piece. Sort of like scaffolding, keeps your house up while you're working on it, though longer term you shouldn't need it.

> But then, we'd like more simplified approach to the solution.
> 
> I'm not sure NEWER/OLDER Future/Lagecy makes life easier for us. I think it will require us to rename this new `Future` to `Legacy` when we add another snapshot version? That's a bit tedious.

There were two considerations in my mind with this -
1. the snapshot format (currently) encompasses a tar archive with a bank / bankrc serialised file, a serialised status cache file, the memory mapped accounts files, etc.; there can be changes in the overall snapshot format that don't affect bankrc serialisation and vice, so I don't think they should be coupled (well at this point at any rate)
2. @mvines seemed quite clear to me on his desire for there to be maximum of two possible versions in use at any time - when there's to be a change in format, the next release will support reading the older version and the newer one, but will only output the newer version; and the next release will drop support for the older version entirely.

Along these lines then the "x.y.z" snapshot version number remains the concern of ledger/src/snapshot_utils.rs, and snapshot_utils then maps this to OLDER or NEWER as appropriate; (note the other serialisations (i.e. status cache, etc.) could follow this approach too).

If a new snapshot format is to be released, the "x.y.z" numbers are updated - one is mapped to NEWER and the previous one to OLDER. In the case of bankrc serialiation, if there's no change there then nothing needs to change in runtime, otherwise NEWER will need to be made OLDER and then a newer NEWER implemented. In the next release after a snapshot version change with a new NEWER, the OLDER is then dropped and is aliased to NEWER.

I was looking ideally for two unordered names but settled on NEWER and OLDER as there should never be any question which version is which and which one you want, as for an upgrade you always want the NEWER one. But maybe this might be more difficult in practice - if so, the convention is easily changed.

However having lots of different kinds of x_y_z constants all over the place that constantly need to be kept in agreement lest the world comes to an end / bad things happen / etc. is asking for trouble in the longer term.

> Instead, I think it's fine to embed explicit specific version numbers in constants. Even at a time, there might temporarily be 3 versions when actively developping, that's fine.

The ser/des flavours in serde_snapshot can be identified be version number, that's also entirely reasonable however those versions do not have a bijective mapping with the snapshot versions, so they should be quite different to ensure there is no chance of confusion.
 
> In addition, we also want to avoid to define like `serialize_iter_as_seq`/`serialize_iter_as_map`/`serialize_iter_as_tuple` by ourselves unless it's absolutely needed.

As I pointed, it is absolutely needed to implement the Serialize trait properly for larger data structures with for instance using serialize_map, serialize_seq, serialize_tuple, serialize_struct, etc. as you otherwise have to first copy your data into a data struct that does implement them - currently account storage entry lists are copied into lists and then into a hash map which is then dumped out - but after a ton of heap allocation, hash calcs, insertions all of which is unnecessary work that takes a lot of time.

Note serde provides collect_map() and collect_seq() - this was in fact the inspiration for the serialize_iter_as_xxx() functions above.
 
> And as [I commented yesterday here](https://github.com/solana-labs/solana/pull/9980#discussion_r427892754), I think there is more simplified approach.
> 
> For example, the code this comment is attached to is serializing AccountsDB. And the corresponding code in my suggested commit is like [this](https://github.com/ryoqun/solana/commit/6d58b73c47294bfb93465d5a83cd2175660b6e6d#diff-90c69309ec8cec7cf741c70e18bf2e86R276).
> 
> I think my suggestion is simpler and more manageable in my opinion.
> 
> Or I might be missing some. Could you explain why you think this pr's current approach is expected to be simper and future-proof?

The big problem is:
```
let entries = serializable_db
    .account_storage_entries
    .iter()
    .map(|x| {
        (
            x.first().unwrap().slot,
            x.iter()
                .map(|x| Self::SerializableAccountStorageEntry::from(x.as_ref()))
                .collect::<Vec<_>>(),
        )
    })
    .collect::<HashMap<Slot, _>>();
```
where account_storage_entries is a list of lists of ASEs, it creates a new vector for each sublist and then shoves them into a hash map until when it's put into a tuple which is then serialized - there is no reason why the ASEs can't or shouldn't be written out to disk as they are processed.

With the serialize_iter_as_map() helper function you can write:
```
let entries = serialize_iter_as_map(
    self.account_storage_entries.iter().map(
        |x| (
            x.first().unwrap().slot,
            serialize_iter_as_seq(
                x.iter()
                    .map(|x| C::SerializableAccountStorageEntry::from(x.as_ref())),
            ),
        )
    )
);
```
and the later in the 4-tuple, include it as `&entries` - this won't create any extra Vecs or HashMaps, and won't do any serizalition until the tuple starts serializing and gets to that element.

(And yes, there's little practical overhead in creating a four element tuple itself at top level and serializing it as it's only four pointers but the data they point to must already exist - or the objects are Serializable which is what serialize_iter_as_xxx() returns, and serialize on the fly).

> Specifically, could you explain how could we reach to [this state](https://github.com/ryoqun/solana/commit/6d58b73c47294bfb93465d5a83cd2175660b6e6d) or even simpler versioned code starting from the current implementation?

Having pulled your branch and looked at it in an editor, the intent is now clearer to me.
Github made everything look green and it looked like all the code had duplicated for each SerdeContext trait implementation (including the context_bankrc/accountsdb_to/from_stream() functions) - actually github elided these because they hadn't changed.

Please accept my apologies.

That bookmarklet script would been very useful yesterday!

> I mean, could you elaborate a bit on this?: [#9980 (comment)](https://github.com/solana-labs/solana/pull/9980#discussion_r428278811)
> 
> > Once the legacy serialisation style is no longer needed, much of this code can and should be removed, as well as the generic type parameterized implementation if it proves to be more trouble than it's worth (I must admit it is quite awkward, especially in rust).

My thoughts with that were to save that for a later PR for the release after the version change, which would remove the *V1_1_0 types, as well as the legacy_or_zero and other trait functions that dealt with the u64 length fields.

> Also, how could we remove the Legacy serialization debt along the course from now, including like `legacy_serialize_byte_length`, which is in the common codepath? Maybe just remove after we remove the current Legacy? If so, why cannot we go one step further and remove them now?

Argument goes likes this, you have to remove the legacy stuff either now or later (after upgrade rolled out). If you wait till later, you just remove the legacy stuff and you're done. If you do it now however, you also have to duplicate some more code to so the legacy format still works. Later after the upgrade is rolled out, you still have some stuff left to delete (but it's easier).

Unless the presence of the legacy stuff is blocking some other development or something else, or if there's a risk that if it doesn't happen now it will be forgotton about, then the wait until later approach requires less total time / resource but still gets you to clean state eventually.

Reply:
> @ryoqun Having pulled down your repo and looked at your commit in an editor (as opposed to github), your intent is much clearer to my now and much of my concern is proven unfounded.

Nice! We're on the same page, then? I've read the rest of comment. I have no doubt that you're very knowledgeable and your code is well reasoned and thought one. I thought code speaks better than words... Seems like I'm too pragmatic coding-first guy. ;)

Then, I hope that this pr's direction is corrected by you. I'm afraid of wasting your time because some code will be reverted. But at least I learnt a lot from your code and comments!

Some extra remarks:

> one is mapped to NEWER and the previous one to OLDER. ... In the next release after a snapshot version change with a new NEWER, the OLDER is then dropped and is aliased to NEWER.

Yeah, I could imagine that. That's neat but we generally prefer less diff/code shifting solution.

> those versions do not have a bijective mapping with the snapshot versions

I think it's fine for bijective mapping to exist between snapshot versions and particular versioned codepath for particular struct like AccountDB/StatusCache. The motto is let's be simple unless needed.

> As I pointed, it is absolutely needed to implement the Serialize trait properly for larger data structures with for instance using serialize_map, serialize_seq, serialize_tuple, serialize_struct, etc. as you otherwise have to first copy your data into a data struct that does implement them...

Yeah, certainly heap copy should be avoided and your solution does it correct in that regard. But, this might not be a problem for now. If there is indeed problem, then we could take your code with comments justifying so. Otherwise, we prefer less code. To be honest, I just didn't notice the wasted memory allocation could happen. Thanks for pointing out!

I dunno rustc is clever enough to elide these copies. I wish so but I guess not. ;)

> Argument goes likes this, you have to remove the legacy stuff either now or later (after upgrade rolled out)...

Here too, your reasoning is correct! But for us, the amount of code diff and the count of prs weights a bit more. Then, you might have guessed this now: Changing code and reviewing diff takes human time. Let's take a path of fewer of them. And our most precious resource is engineering time and we can't increase the throughput of it easily. So the only way for more productivity is to reduce the traffic (code changes) ;)

Reply:
Yeah the stuff you picked out into the versioned traits was a lot more targeted / limited than what it looked like at first in github's diff view - it makes sense.

The PR branch is merged with your stuff, rebased off of master and is close, but not quite, to where it needs to be. The older / newer enum is still there but fixing that up is cosmetic.

One aspect of serde that I've come to appreciate is it doesn't really matter what type you think your data is, what matters is what serde thinks it is - and it can only ever be one of the 29 serde types listed here `https://serde.rs/data-model.html`. That said, some of the serializers can't tell the difference between some of the type (i.e. bincode appears unable to distinguish between tuples and tuple structs for instance) whereas using an XML serializer (say) with serde should embed a lot more metadata in the serialized output. Actually this could make for an interesting test case....

Reply:
@svenski123 Whoo! Thanking for quickly reworking on this.

> The PR branch is merged with your stuff, rebased off of master and is close, but not quite, to where it needs to be. The older / newer enum is still there but fixing that up is cosmetic.

Yeah, those is cosmetics. We can fix it later.

Congratulation! I think this pr is ready for merge. I'm soo appreciating your hard work :)

Reply:
> One aspect of serde that I've come to appreciate is it doesn't really matter what type you think your data is, what matters is what serde thinks it is - and it can only ever be one of the 29 serde types listed here

Hehe, that sounds interesting observation. Could you tell me what did enlighten this insight from this PR? Maybe some relaxed coercion from variants of rust types into one of serde types? Which one, specifically?

> said, some of the serializers can't tell the difference between some of the type ...

Yeah, that's true. But because of the compromise, serde established the role of marketplace between application serializer logics and actual encoder logics in the rust ecosystem. This is much like Pandoc and I think they done decent job for it with fine balance for common denominator. For us as performance-oriented project, it's a bit bloated, though... ;) serde's primary target is web backends, after all.